### PR TITLE
Documentation on how to implement a new service

### PR DESCRIPTION
**Pull request by DarkaMaul**
Following the last outage due to #16302, I wrote some documentation on how a new service should be added within `warehouse`.

This may be useful for the next new developers (or serve as a review baseline). 

/cc @woodruffw @facutuesca

### CODE REVIEW

File: docs/dev/development/patterns.rst [hunk 1]
```diff
@@ -180,9 +180,7 @@ Testing the service itself
 1. Implement a ``test_includeme`` function to test the service registration.
 2. Test each service implementation individually to meet ``warehouse`` 100% test coverage.
 
-   - Write a ``Test<ServiceName>`` class and implement a ``test_interface_matches`` function (the
-   exact name is irrelevant) to verify that the service implementation matches the interface definition
-   using the ``verifyClass`` function from zope.
+   - Write a ``Test<ServiceName>`` class and implement ``test_interface_matches`` function (the exact name is irrelevant) to verify that the service implementation matches the interface definition using the ``verifyClass`` function from zope.
```
Comment (woodruffw) [inline docs/dev/development/patterns.rst:183]: I think this parenthetical is implied by the context:

```suggestion
   - Write a ``Test<ServiceName>`` class and implement ``test_interface_matches`` function to verify that the service implementation matches the interface definition using the ``verifyClass`` function from zope.
```

Maybe it'd be good to link to a source example of this, though?

File: docs/dev/development/patterns.rst [hunk 2]
```diff
@@ -107,6 +107,107 @@ Class                     Method
...
+   Any new implementation must implement the complete interface, including all its methods and attributes.
+
+4. Implement each service creation method. If the Service is simple enough, use a class method in
+   your service implementation (usually named ``create_service``). For more complex cases, implement
+   a ``ServiceFactory`` class, responsible to create the service instance.
+
+5. Register the service. The new service(s) must be registered to be available in the request object.
+
+   - If you have multiple services, create a new setting (in ``warehouse/config.py``) to select
+     which backend to use.
```
Comment (di) [inline docs/dev/development/patterns.rst:145]: Probably should be something like "if there are multiple variations of the same service" or something like that.

**Review by di — APPROVED** [review]
LGTM aside from outstanding comments

File: docs/dev/development/patterns.rst [hunk 3]
```diff
@@ -107,6 +107,107 @@ Class                     Method
...
+When adding new services to ``warehouse``, the following checklist serves as a comprehensive guideline to ensure
+you stay on track.
+
+Adding a new service
+~~~~~~~~~~~~~~~~~~~~~
+
+1. Create an Interface for the service. The interface serves as the baseline of the new service (design by
+   contract pattern) and details all methods and attributes shared by the different service implementations.
+
+   Warehouse uses zope_ to define interfaces. The interfaces are usually declared in a file named
```
Comment (miketheman) [inline docs/dev/development/patterns.rst:126]: nitpick: the package is specifically `zope.interface` vs zope, the framework.

File: docs/dev/development/patterns.rst [hunk 4]
```diff
@@ -107,6 +107,107 @@ Class                     Method
...
+   ``interfaces.py`` within the relevant component, such as ``packaging/interfaces.py``.
+
+2. Create the new service. The service must define all methods and attributes declared in the interface.
+   This implementation contains the core logic of the service features.
```
Comment (miketheman) [inline docs/dev/development/patterns.rst:130]: suggestion: services MAY add further methods specific to the service that are not required on all implementations of the interface.

File: docs/dev/development/patterns.rst [hunk 5]
```diff
@@ -107,6 +107,107 @@ Class                     Method
 ``HTTPServerError`` (5xx) Raise
 ========================= ==================================
 
+Implementing new services
+=========================
+
+Warehouse uses services to provide pluggable functionalities within the codebase. They are implemented using
+`pyramid-service`_. After being registered, services are accessible using the ``find_service`` method of the
```
Comment (miketheman) [inline docs/dev/development/patterns.rst:114]: nit: the project is named `pyramid-services` (plural)

File: docs/dev/development/patterns.rst [hunk 6]
```diff
@@ -107,6 +107,107 @@ Class                     Method
...
+     function for Pyramid to detect it and use the service factory created at the previous step.
+
+6. (Optional) Add the new module to the ``warehouse/config.py``. If the new service is defined in a
+   new module, add the new module within the warehouse ``configure`` function. This enrollment
+   ensures Pyramid can detect it.
+
+Using the service
+~~~~~~~~~~~~~~~~~
+
+To use a service, query it using ``request.find_services`` with the service interface. This
```
Comment (miketheman) [inline docs/dev/development/patterns.rst:161]:
```suggestion
To use a service, query it using ``request.find_service()`` with the service interface. This
```

File: docs/dev/development/patterns.rst [hunk 7]
```diff
@@ -107,6 +107,107 @@ Class                     Method
...
+method will return an instance of the service correctly selected based on the context and environment.
+
+Example:
+
+.. code-block:: python
+
+   metrics = request.find_service(IMetricsService, context=None)
```
Comment (miketheman) [inline docs/dev/development/patterns.rst:168]: thought: `context=` is not always required/used, nor is `name=` - would it be better to show an example without `context=None`, or have examples for using named service implementations?

**Review by miketheman — CHANGES_REQUESTED** [review]
Nice work, had some notes inline.

**Review by miketheman — APPROVED** [review]

**Pull request merged**